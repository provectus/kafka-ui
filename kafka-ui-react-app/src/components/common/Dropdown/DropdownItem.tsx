import React, { PropsWithChildren } from 'react';
import { ClickEvent, MenuItem, MenuItemProps } from '@szhsin/react-menu';
import { useConfirm } from 'lib/hooks/useConfirm';

import * as S from './Dropdown.styled';

export interface DropdownItemProps extends PropsWithChildren<MenuItemProps> {
  danger?: boolean;
  onClick?(): void;
  confirm?: React.ReactNode;
}

const DropdownItem = React.forwardRef<unknown, DropdownItemProps>(
  ({ onClick, danger, children, confirm, ...rest }, ref) => {
    const confirmation = useConfirm();

    const handleClick = (e: ClickEvent) => {
      if (!onClick) return;

      // eslint-disable-next-line no-param-reassign
      e.stopPropagation = true;
      e.syntheticEvent.stopPropagation();

      if (confirm) {
        confirmation(confirm, onClick);
      } else {
        onClick();
      }
    };

    return (
      <MenuItem onClick={handleClick} {...rest} ref={ref}>
        {danger ? <S.DangerItem>{children}</S.DangerItem> : children}
      </MenuItem>
    );
  }
);

export default DropdownItem;
