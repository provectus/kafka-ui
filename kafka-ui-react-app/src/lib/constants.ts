import { SelectOption } from 'components/common/Select/Select';
import { ConfigurationParameters } from 'generated-sources';

declare global {
  interface Window {
    basePath: string;
  }
}

export const BASE_PARAMS: ConfigurationParameters = {
  basePath: window.basePath || '',
  credentials: 'include',
  headers: {
    'Content-Type': 'application/json',
  },
};

export const TOPIC_NAME_VALIDATION_PATTERN = /^[a-zA-Z0-9._-]+$/;
export const SCHEMA_NAME_VALIDATION_PATTERN = /^[.,A-Za-z0-9_/-]+$/;

export const TOPIC_CUSTOM_PARAMS_PREFIX = 'customParams';
export const TOPIC_CUSTOM_PARAMS: Record<string, string> = {
  'compression.type': 'producer',
  'leader.replication.throttled.replicas': '',
  'message.downconversion.enable': 'true',
  'segment.jitter.ms': '0',
  'flush.ms': '9223372036854775807',
  'follower.replication.throttled.replicas': '',
  'segment.bytes': '1073741824',
  'flush.messages': '9223372036854775807',
  'message.format.version': '2.3-IV1',
  'file.delete.delay.ms': '60000',
  'max.compaction.lag.ms': '9223372036854775807',
  'min.compaction.lag.ms': '0',
  'message.timestamp.type': 'CreateTime',
  preallocate: 'false',
  'min.cleanable.dirty.ratio': '0.5',
  'index.interval.bytes': '4096',
  'unclean.leader.election.enable': 'true',
  'retention.bytes': '-1',
  'delete.retention.ms': '86400000',
  'segment.ms': '604800000',
  'message.timestamp.difference.max.ms': '9223372036854775807',
  'segment.index.bytes': '10485760',
};

export const MILLISECONDS_IN_WEEK = 604_800_000;
export const MILLISECONDS_IN_DAY = 86_400_000;
export const MILLISECONDS_IN_SECOND = 1_000;

export const NOT_SET = -1;
export const BYTES_IN_GB = 1_073_741_824;
export const BUILD_VERSION_PATTERN = /v\d.\d.\d/;

export const PER_PAGE = 25;
export const MESSAGES_PER_PAGE = '100';

export const GIT_REPO_LINK = 'https://github.com/provectus/kafka-ui';
export const GIT_REPO_LATEST_RELEASE_LINK =
  'https://api.github.com/repos/provectus/kafka-ui/releases/latest';

export const LOCAL_STORAGE_KEY_PREFIX = 'kafka-ui';

export enum AsyncRequestStatus {
  initial = 'initial',
  pending = 'pending',
  fulfilled = 'fulfilled',
  rejected = 'rejected',
}

export const QUERY_REFETCH_OFF_OPTIONS = {
  refetchOnMount: false,
  refetchOnWindowFocus: false,
  refetchIntervalInBackground: false,
};

// Cluster Form Constants
export const AUTH_OPTIONS: SelectOption[] = [
  { value: 'SASL/JAAS', label: 'SASL/JAAS' },
  { value: 'SASL/GSSAPI', label: 'SASL/GSSAPI' },
  { value: 'SASL/OAUTHBEARER', label: 'SASL/OAUTHBEARER' },
  { value: 'SASL/PLAIN', label: 'SASL/PLAIN' },
  { value: 'SASL/SCRAM-256', label: 'SASL/SCRAM-256' },
  { value: 'SASL/SCRAM-512', label: 'SASL/SCRAM-512' },
  { value: 'Delegation tokens', label: 'Delegation tokens' },
  { value: 'SASL/LDAP', label: 'SASL/LDAP' },
  { value: 'SASL/AWS IAM', label: 'SASL/AWS IAM' },
  { value: 'mTLS', label: 'mTLS' },
];

export const SECURITY_PROTOCOL_OPTIONS: SelectOption[] = [
  { value: 'SASL_SSL', label: 'SASL_SSL' },
  { value: 'SASL_PLAINTEXT', label: 'SASL_PLAINTEXT' },
];
export const METRICS_OPTIONS: SelectOption[] = [
  { value: 'JMX', label: 'JMX' },
  { value: 'PROMETHEUS', label: 'PROMETHEUS' },
];
