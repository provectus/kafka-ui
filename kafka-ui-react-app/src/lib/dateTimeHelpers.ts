export const formatTimestamp = (
  timestamp?: number | string | Date,
  format: Intl.DateTimeFormatOptions = { hour12: false }
): string => {
  if (!timestamp) {
    return '';
  }

  // empty array gets the default one from the browser
  const date = new Date(timestamp);

  // invalid date
  if (Number.isNaN(date.getTime())) {
    return '';
  }

  // browser support
  const language = navigator.language || navigator.languages[0];
  return date.toLocaleString(language || [], format);
};

export const formatMilliseconds = (input = 0) => {
  const milliseconds = Math.max(input || 0, 0);

  const seconds = Math.floor(milliseconds / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);

  if (hours > 0) {
    return `${hours}h ${minutes % 60}m`;
  }

  if (minutes > 0) {
    return `${minutes}m ${seconds % 60}s`;
  }

  if (seconds > 0) {
    return `${seconds}s`;
  }

  return `${milliseconds}ms`;
};

export const passedTime = (value: number) => (value < 10 ? `0${value}` : value);

export const calculateTimer = (startedAt: number) => {
  const nowDate = new Date();
  const now = nowDate.getTime();
  const newDate = now - startedAt;
  const minutes = nowDate.getMinutes();
  const second = nowDate.getSeconds();

  return newDate > 0 ? `${passedTime(minutes)}:${passedTime(second)}` : '00:00';
};
