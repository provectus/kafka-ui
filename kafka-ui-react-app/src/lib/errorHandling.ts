import { ServerResponse } from 'redux/interfaces';

export const getResponse = async (
  response: Response
): Promise<ServerResponse> => {
  let body;
  try {
    body = await response.json();
  } catch (e) {
    // do nothing;
  }
  return {
    status: response.status,
    statusText: response.statusText,
    url: response.url,
    message: body?.message,
  };
};
