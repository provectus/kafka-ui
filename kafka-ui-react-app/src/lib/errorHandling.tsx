import React from 'react';
import Alert from 'components/common/Alert/Alert';
import toast, { ToastType } from 'react-hot-toast';
import { ErrorResponse } from 'generated-sources';

interface ServerResponse {
  status: number;
  statusText: string;
  url?: string;
  message?: ErrorResponse['message'];
}
export type ToastTypes = ToastType | 'warning';

export const getResponse = async (
  response: Response
): Promise<ServerResponse> => {
  let body;
  try {
    body = await response.json();
  } catch (e) {
    // do nothing;
  }
  return {
    status: response.status,
    statusText: response.statusText,
    url: response.url,
    message: body?.message,
  };
};

interface AlertOptions {
  id?: string;
  title?: string;
  message: React.ReactNode;
}

export const showAlert = (
  type: ToastTypes,
  { title, message, id }: AlertOptions
) => {
  toast.custom(
    (t) => (
      <Alert
        title={title || ''}
        type={type}
        message={message}
        onDissmiss={() => toast.remove(t.id)}
      />
    ),
    { id }
  );
};

export const showSuccessAlert = (options: AlertOptions) => {
  showAlert('success', {
    ...options,
    title: options.title || 'Success',
  });
};

export const showServerError = async (
  response: Response,
  options?: AlertOptions
) => {
  let body: Record<string, string> = {};
  try {
    body = await response.json();
  } catch (e) {
    // do nothing;
  }
  if (response.status) {
    showAlert('error', {
      id: response.url,
      title: `${response.status} ${response.statusText}`,
      message: body?.message || 'An error occurred',
      ...options,
    });
  } else {
    showAlert('error', {
      id: 'server-error',
      title: `Something went wrong`,
      message: 'An error occurred',
      ...options,
    });
  }
};
