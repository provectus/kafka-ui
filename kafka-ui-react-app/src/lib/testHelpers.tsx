import React, { PropsWithChildren, ReactElement, useMemo } from 'react';
import {
  MemoryRouter,
  MemoryRouterProps,
  Route,
  Routes,
} from 'react-router-dom';
import fetchMock from 'fetch-mock';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { theme } from 'theme/theme';
import {
  render,
  renderHook,
  RenderOptions,
  waitFor,
} from '@testing-library/react';
import { AnyAction, Store } from 'redux';
import { RootState } from 'redux/interfaces';
import { configureStore } from '@reduxjs/toolkit';
import rootReducer from 'redux/reducers';
import {
  QueryClient,
  QueryClientProvider,
  UseQueryResult,
} from '@tanstack/react-query';
import { ConfirmContextProvider } from 'components/contexts/ConfirmContext';
import ConfirmationModal from 'components/common/ConfirmationModal/ConfirmationModal';
import { GlobalSettingsContext } from 'components/contexts/GlobalSettingsContext';
import { UserInfoRolesAccessContext } from 'components/contexts/UserInfoRolesAccessContext';

import { RolesType, modifyRolesData } from './permissions';

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  preloadedState?: Partial<RootState>;
  store?: Store<Partial<RootState>, AnyAction>;
  initialEntries?: MemoryRouterProps['initialEntries'];
  userInfo?: {
    roles?: RolesType;
    rbacFlag: boolean;
  };
}

interface WithRouteProps {
  children: React.ReactNode;
  path: string;
}

export const expectQueryWorks = async (
  mock: fetchMock.FetchMockStatic,
  result: { current: UseQueryResult<unknown, unknown> }
) => {
  await waitFor(() => expect(result.current.isFetched).toBeTruthy());
  expect(mock.calls()).toHaveLength(1);
  expect(result.current.data).toBeDefined();
};

export const WithRoute: React.FC<WithRouteProps> = ({ children, path }) => {
  return (
    <Routes>
      <Route path={path} element={children} />
    </Routes>
  );
};

export const TestQueryClientProvider: React.FC<PropsWithChildren<unknown>> = ({
  children,
}) => {
  // use new QueryClient instance for each test run to avoid issues with cache
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });
  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};

/**
 * @description it will create a UserInfo Provider that will actually
 * disable the rbacFlag , to user if you can pass it as an argument
 * */
const TestUserInfoProvider: React.FC<
  PropsWithChildren<{ data?: { roles?: RolesType; rbacFlag: boolean } }>
> = ({ children, data }) => {
  const contextValue = useMemo(() => {
    const roles = modifyRolesData(data?.roles);

    return {
      username: 'test',
      rbacFlag: !!(typeof data?.rbacFlag === 'undefined'
        ? false
        : data?.rbacFlag),
      roles,
    };
  }, [data]);

  return (
    <UserInfoRolesAccessContext.Provider value={contextValue}>
      {children}
    </UserInfoRolesAccessContext.Provider>
  );
};

const customRender = (
  ui: ReactElement,
  {
    preloadedState,
    store = configureStore<RootState>({
      reducer: rootReducer,
      preloadedState,
    }),
    initialEntries,
    userInfo,
    ...renderOptions
  }: CustomRenderOptions = {}
) => {
  // overrides @testing-library/react render.
  const AllTheProviders: React.FC<PropsWithChildren<unknown>> = ({
    children,
  }) => (
    <TestQueryClientProvider>
      <GlobalSettingsContext.Provider value={{ hasDynamicConfig: false }}>
        <ThemeProvider theme={theme}>
          <TestUserInfoProvider data={userInfo}>
            <ConfirmContextProvider>
              <Provider store={store}>
                <MemoryRouter initialEntries={initialEntries}>
                  <div>
                    {children}
                    <ConfirmationModal />
                  </div>
                </MemoryRouter>
              </Provider>
            </ConfirmContextProvider>
          </TestUserInfoProvider>
        </ThemeProvider>
      </GlobalSettingsContext.Provider>
    </TestQueryClientProvider>
  );
  return render(ui, { wrapper: AllTheProviders, ...renderOptions });
};

const customRenderHook = (hook: () => UseQueryResult<unknown, unknown>) =>
  renderHook(hook, { wrapper: TestQueryClientProvider });

export { customRender as render, customRenderHook as renderQueryHook };

export class EventSourceMock {
  url: string;

  close: () => void;

  open: () => void;

  error: () => void;

  onmessage: () => void;

  constructor(url: string) {
    this.url = url;
    this.open = jest.fn();
    this.error = jest.fn();
    this.onmessage = jest.fn();
    this.close = jest.fn();
  }
}
