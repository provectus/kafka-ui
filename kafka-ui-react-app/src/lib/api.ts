import {
  KsqlApi,
  TopicsApi,
  SchemasApi,
  BrokersApi,
  MessagesApi,
  ClustersApi,
  Configuration,
  KafkaConnectApi,
  ConsumerGroupsApi,
  AuthorizationApi,
  ApplicationConfigApi,
} from 'generated-sources';
import { BASE_PARAMS } from 'lib/constants';

const apiClientConf = new Configuration(BASE_PARAMS);

export const ksqlDbApiClient = new KsqlApi(apiClientConf);
export const topicsApiClient = new TopicsApi(apiClientConf);
export const brokersApiClient = new BrokersApi(apiClientConf);
export const schemasApiClient = new SchemasApi(apiClientConf);
export const messagesApiClient = new MessagesApi(apiClientConf);
export const clustersApiClient = new ClustersApi(apiClientConf);
export const kafkaConnectApiClient = new KafkaConnectApi(apiClientConf);
export const consumerGroupsApiClient = new ConsumerGroupsApi(apiClientConf);
export const authApiClient = new AuthorizationApi(apiClientConf);
export const appConfigApiClient = new ApplicationConfigApi(apiClientConf);
