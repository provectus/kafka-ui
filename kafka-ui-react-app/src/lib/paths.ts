import { Broker, Connect, Connector } from 'generated-sources';
import { ClusterName, SchemaName, TopicName } from 'redux/interfaces';

import { GIT_REPO_LINK } from './constants';
import { ConsumerGroupID } from './hooks/api/consumers';

export const gitCommitPath = (commit: string) =>
  `${GIT_REPO_LINK}/commit/${commit}`;

export enum RouteParams {
  clusterName = ':clusterName',
  consumerGroupID = ':consumerGroupID',
  subject = ':subject',
  topicName = ':topicName',
  connectName = ':connectName',
  connectorName = ':connectorName',
  brokerId = ':brokerId',
}

export const getNonExactPath = (path: string) => `${path}/*`;

export const errorPage = '/404';
export const accessErrorPage = '/403';

export const clusterPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `/ui/clusters/${clusterName}`;

export type ClusterNameRoute = { clusterName: ClusterName };

// Brokers
export const clusterBrokerRelativePath = 'brokers';
export const clusterBrokerMetricsRelativePath = 'metrics';
export const clusterBrokerConfigsRelativePath = 'configs';

export const clusterBrokersPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterPath(clusterName)}/${clusterBrokerRelativePath}`;

export const clusterBrokerPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  brokerId: Broker['id'] | string = RouteParams.brokerId
) => `${clusterBrokersPath(clusterName)}/${brokerId}`;
export const clusterBrokerMetricsPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  brokerId: Broker['id'] | string = RouteParams.brokerId
) =>
  `${clusterBrokerPath(
    clusterName,
    brokerId
  )}/${clusterBrokerMetricsRelativePath}`;

export const clusterBrokerConfigsPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  brokerId: Broker['id'] | string = RouteParams.brokerId
) =>
  `${clusterBrokerPath(
    clusterName,
    brokerId
  )}/${clusterBrokerConfigsRelativePath}`;

export type ClusterBrokerParam = {
  clusterName: ClusterName;
  brokerId: string;
};

// Consumer Groups
export const clusterConsumerGroupsRelativePath = 'consumer-groups';
export const clusterConsumerGroupResetRelativePath = 'reset-offsets';
export const clusterConsumerGroupResetOffsetsRelativePath = `${RouteParams.consumerGroupID}/${clusterConsumerGroupResetRelativePath}`;
export const clusterConsumerGroupsPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterPath(clusterName)}/${clusterConsumerGroupsRelativePath}`;
export const clusterConsumerGroupDetailsPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  groupId: string = RouteParams.consumerGroupID
) => `${clusterConsumerGroupsPath(clusterName)}/${groupId}`;
export const clusterConsumerGroupResetOffsetsPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  groupId: string = RouteParams.consumerGroupID
) =>
  `${clusterConsumerGroupDetailsPath(
    clusterName,
    groupId
  )}/${clusterConsumerGroupResetRelativePath}`;
export type ClusterGroupParam = {
  consumerGroupID: ConsumerGroupID;
  clusterName: ClusterName;
};

// Schemas
export const clusterSchemasRelativePath = 'schemas';
export const clusterSchemaNewRelativePath = 'create-new';
export const clusterSchemaEditPageRelativePath = `edit`;
export const clusterSchemaSchemaComparePageRelativePath = `compare`;
export const clusterSchemaEditRelativePath = `${RouteParams.subject}/${clusterSchemaEditPageRelativePath}`;
export const clusterSchemaSchemaDiffRelativePath = `${RouteParams.subject}/${clusterSchemaSchemaComparePageRelativePath}`;
export const clusterSchemasPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterPath(clusterName)}/schemas`;
export const clusterSchemaNewPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterSchemasPath(clusterName)}/${clusterSchemaNewRelativePath}`;
export const clusterSchemaPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  subject: SchemaName = RouteParams.subject
) => {
  let subjectName = subject;
  if (subject !== ':subject') subjectName = encodeURIComponent(subject);
  return `${clusterSchemasPath(clusterName)}/${subjectName}`;
};
export const clusterSchemaEditPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  subject: SchemaName = RouteParams.subject
) => {
  let subjectName = subject;
  if (subject !== ':subject') subjectName = encodeURIComponent(subject);
  return `${clusterSchemasPath(clusterName)}/${subjectName}/edit`;
};
export const clusterSchemaComparePath = (
  clusterName: ClusterName = RouteParams.clusterName,
  subject: SchemaName = RouteParams.subject
) => `${clusterSchemaPath(clusterName, subject)}/compare`;

export type ClusterSubjectParam = {
  subject: string;
  clusterName: ClusterName;
};

// Topics
export const clusterTopicsRelativePath = 'all-topics';
export const clusterTopicNewRelativePath = 'create-new-topic';
export const clusterTopicCopyRelativePath = 'copy';
export const clusterTopicsPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterPath(clusterName)}/${clusterTopicsRelativePath}`;
export const clusterTopicNewPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterTopicsPath(clusterName)}/${clusterTopicNewRelativePath}`;
export const clusterTopicCopyPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterTopicsPath(clusterName)}/${clusterTopicCopyRelativePath}`;

// Topics topic
export const clusterTopicSettingsRelativePath = 'settings';
export const clusterTopicMessagesRelativePath = 'messages';
export const clusterTopicConsumerGroupsRelativePath = 'consumer-groups';
export const clusterTopicStatisticsRelativePath = 'statistics';
export const clusterTopicEditRelativePath = 'edit';
export const clusterTopicPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  topicName: TopicName = RouteParams.topicName
) => `${clusterTopicsPath(clusterName)}/${topicName}`;
export const clusterTopicSettingsPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  topicName: TopicName = RouteParams.topicName
) =>
  `${clusterTopicPath(
    clusterName,
    topicName
  )}/${clusterTopicSettingsRelativePath}`;
export const clusterTopicMessagesPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  topicName: TopicName = RouteParams.topicName
) =>
  `${clusterTopicPath(
    clusterName,
    topicName
  )}/${clusterTopicMessagesRelativePath}`;
export const clusterTopicEditPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  topicName: TopicName = RouteParams.topicName
) =>
  `${clusterTopicPath(clusterName, topicName)}/${clusterTopicEditRelativePath}`;
export const clusterTopicConsumerGroupsPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  topicName: TopicName = RouteParams.topicName
) =>
  `${clusterTopicPath(
    clusterName,
    topicName
  )}/${clusterTopicConsumerGroupsRelativePath}`;
export const clusterTopicStatisticsPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  topicName: TopicName = RouteParams.topicName
) =>
  `${clusterTopicPath(
    clusterName,
    topicName
  )}/${clusterTopicStatisticsRelativePath}`;

export type RouteParamsClusterTopic = {
  clusterName: ClusterName;
  topicName: TopicName;
};

// Kafka Connect
export const clusterConnectsRelativePath = 'connects';
export const clusterConnectorsRelativePath = 'connectors';
export const clusterConnectorNewRelativePath = 'create-new';
export const clusterConnectConnectorsRelativePath = `${RouteParams.connectName}/connectors`;
export const clusterConnectConnectorRelativePath = `${clusterConnectConnectorsRelativePath}/${RouteParams.connectorName}`;
const clusterConnectConnectorTasksRelativePath = 'tasks';
export const clusterConnectConnectorConfigRelativePath = 'config';

export const clusterConnectsPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterPath(clusterName)}/connects`;
export const clusterConnectorsPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterPath(clusterName)}/connectors`;
export const clusterConnectorNewPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterConnectorsPath(clusterName)}/create-new`;
export const clusterConnectConnectorsPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  connectName: Connect['name'] = RouteParams.connectName
) => `${clusterConnectsPath(clusterName)}/${connectName}/connectors`;
export const clusterConnectConnectorPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  connectName: Connect['name'] = RouteParams.connectName,
  connectorName: Connector['name'] = RouteParams.connectorName
) =>
  `${clusterConnectConnectorsPath(clusterName, connectName)}/${connectorName}`;
export const clusterConnectConnectorEditPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  connectName: Connect['name'] = RouteParams.connectName,
  connectorName: Connector['name'] = RouteParams.connectorName
) =>
  `${clusterConnectConnectorsPath(
    clusterName,
    connectName
  )}/${connectorName}/edit`;
export const clusterConnectConnectorTasksPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  connectName: Connect['name'] = RouteParams.connectName,
  connectorName: Connector['name'] = RouteParams.connectorName
) =>
  `${clusterConnectConnectorPath(
    clusterName,
    connectName,
    connectorName
  )}/${clusterConnectConnectorTasksRelativePath}`;
export const clusterConnectConnectorConfigPath = (
  clusterName: ClusterName = RouteParams.clusterName,
  connectName: Connect['name'] = RouteParams.connectName,
  connectorName: Connector['name'] = RouteParams.connectorName
) =>
  `${clusterConnectConnectorPath(
    clusterName,
    connectName,
    connectorName
  )}/${clusterConnectConnectorConfigRelativePath}`;

export type RouterParamsClusterConnectConnector = {
  clusterName: ClusterName;
  connectName: Connect['name'];
  connectorName: Connector['name'];
};

// KsqlDb
export const clusterKsqlDbRelativePath = 'ksqldb';
export const clusterKsqlDbQueryRelativePath = 'query';
export const clusterKsqlDbTablesRelativePath = 'tables';
export const clusterKsqlDbStreamsRelativePath = 'streams';

export const clusterKsqlDbPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterPath(clusterName)}/${clusterKsqlDbRelativePath}`;
export const clusterKsqlDbQueryPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterKsqlDbPath(clusterName)}/${clusterKsqlDbQueryRelativePath}`;
export const clusterKsqlDbTablesPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterKsqlDbPath(clusterName)}/${clusterKsqlDbTablesRelativePath}`;
export const clusterKsqlDbStreamsPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterKsqlDbPath(clusterName)}/${clusterKsqlDbStreamsRelativePath}`;

// Cluster Config
export const clusterConfigRelativePath = 'config';
export const clusterConfigPath = (
  clusterName: ClusterName = RouteParams.clusterName
) => `${clusterPath(clusterName)}/${clusterConfigRelativePath}`;

const clusterNewConfigRelativePath = 'create-new-cluster';
export const clusterNewConfigPath = `/ui/clusters/${clusterNewConfigRelativePath}`;
