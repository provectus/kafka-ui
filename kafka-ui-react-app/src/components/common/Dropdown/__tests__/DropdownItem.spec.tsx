import React from 'react';
import DropdownItem from 'components/common/Dropdown/DropdownItem';
import { render } from 'lib/testHelpers';
import userEvent from '@testing-library/user-event';
import { screen } from '@testing-library/react';

const onClick = jest.fn();

describe('DropdownItem', () => {
  it('matches snapshot', () => {
    const { baseElement } = render(
      <DropdownItem onClick={jest.fn()}>Item 1</DropdownItem>
    );
    expect(onClick).not.toHaveBeenCalled();
    expect(baseElement).toBeInTheDocument();
  });

  it('handles Click', () => {
    render(<DropdownItem onClick={onClick}>Item 1</DropdownItem>);

    const dropDown = screen.getByText('Item 1');

    userEvent.click(dropDown);
    expect(onClick).toHaveBeenCalled();
  });
});
