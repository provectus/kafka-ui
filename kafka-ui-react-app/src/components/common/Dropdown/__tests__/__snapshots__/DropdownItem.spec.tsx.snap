// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownItem matches snapshot 1`] = `
.c0 {
  color: initial;
}

<DropdownItem
  onClick={[MockFunction]}
>
  <styled.a
    $isDanger={false}
    className="dropdown-item is-link"
    onClick={[Function]}
  >
    <a
      className="c0 dropdown-item is-link"
      href="#end"
      onClick={[Function]}
      role="menuitem"
      type="button"
    >
      Item 1
    </a>
  </styled.a>
</DropdownItem>
`;
